### ISSUE TITLE
Cannot play Midi file from WPS

### ISSUE THREAD

**Issue by lerdmann** [komh/ksoftseq#9]
Using: flac-1.3.4, fluidsynth-libs-2.1.5, flac-libs-1.3.4, kmididec-0.3.1, ksoftseq-1.1.2. I also installed timidity (command line).

Playing a MIDI file from the WPS is unstable. What does work is to play the MIDI file via either kmidi.exe or kmidimmio.exe (from the kmididec package) or to play the file via timidity.exe.

Is ksoftseq.dll (the MCD driver) designed to handle multiple instances ? I am asking as I remember from the past that the WPS will attempt to start 2 instances of the MCD (why, I don't know) when you either double click on a MIDI file or if you attempt to open a sound file via MPPM.EXE and if the MCD does not support multiple instances, then the WPS will run into trouble. The same was true for Timidity MCD and the problem started to pop up when computers started to have more than one core.

By the way: I looked into ksoftseq.dll and I could not find any load time linking to any sound related DLL but apparently you are run time linking to libkai.dll and fluidsy2.dll , is that correct ? But I could not find any reference to kmididec. Is it necessary to install kmididec for ksoftseq to work ?

**Comment by komh** [comment]
Hi/2.

What do you mean 'unstable'?

ksoftseq is designed to support multiple instances. And here ksoftseq works well on WPS and with MPPM.EXE.

ksoftseq is a standalone DLL which does not depend on any other user DLLs. Why do you think ksoftseq is linking to libkai.dll and fluidsy2.dll ?

Anyway here is the latest test version of ksoftseq.dll. Test please...

[ksoftseq-test.zip](https://github.com/komh/ksoftseq/files/11935506/ksoftseq-test.zip)

**Comment by lerdmann** [comment]
"playing is unstable": when I change volume during playing, the sound will go into a loop. This happens every time I try. I do not have this problem when I play an ordinary WAV or MP3 file.

**Comment by lerdmann** [comment]
tried kostseq-test.zip. Same problem, changing sound volume makes sound go into a loop.

**Comment by komh** [comment]
I cannot reproduce this. So I have questions.

1. What are steps to play a MIDI file on WPS ?
2. How do you change the volume while playing ?
3. What are your env. vars. for KAI which you can get with ```set | grep KAI``` ?

**Comment by lerdmann** [comment]
1. I double click a MIDI file, I also used MPPM.exe to open a Midi File and have it play
2. I am using the master volume control (volume.exe)
3. KAI_AUTOMODE=DART

However, need to correct myself: if I use USB audio (USBAUDIO.SYS, USBAUD2.SYS) it just plays fine, there is no looping and the volume control works as expected. I only have problems if I switch over to UNIAUD. Therefore I suspect this to be a problem with UNIAUD and not with ksoftseq.
There is another funny thing: ksoftseq seems to have some problems in finding the default audio device (unless this problem is also UNIAUD related): If I play via USB first and then, I switch the default device (I am delivering a utility with the USB audio drivers that does that and AN has also their own solution to switch the default audio device), I cannot switch to UNIAUD: if I do, ksoftseq will still continue to play on the USB device, if I remove the USB audio device, ksoftseq will refuse to play even if have switched over the default audio device to UNIAUD. Might also be yet another UNIAUD problem.

**Comment by lerdmann** [comment]
Can you tell me what UNIAUD version (UNIAUD16.SYS, UNIAUD32.SYS) you used when testing ? I`ll try and use the same if you can let me know where you got it from.

**Comment by komh** [comment]
> 1. I double click a MIDI file, I also used MPPM.exe to open a Midi File and have it play
> 
>     2. I am using the master volume control (volume.exe)
> 
>     3. KAI_AUTOMODE=DART
> 
> 

Maybe are you running kaisrv.exe ?

> However, need to correct myself: if I use USB audio (USBAUDIO.SYS, USBAUD2.SYS) it just plays fine, there is no looping and the volume control works as expected. I only have problems if I switch over to UNIAUD. Therefore I suspect this to be a problem with UNIAUD and not with ksoftseq. There is another funny thing: ksoftseq seems to have some problems in finding the default audio device (unless this problem is also UNIAUD related): If I play via USB first and then, I switch the default device (I am delivering a utility with the USB audio drivers that does that and AN has also their own solution to switch the default audio device), I cannot switch to UNIAUD: if I do, ksoftseq will still continue to play on the USB device, if I remove the USB audio device, ksoftseq will refuse to play even if have switched over the default audio device to UNIAUD. Might also be yet another UNIAUD problem.

Oooops... This is because ksoftseq initializes kai only once at DLL load time which determines a default audio device. I'll try to fix this.

BTW, would you mind opening a separate ticket for this ?

**Comment by komh** [comment]
> Can you tell me what UNIAUD version (UNIAUD16.SYS, UNIAUD32.SYS) you used when testing ? I`ll try and use the same if you can let me know where you got it from.

I attach my uniaud drivers. 

[uniaud.zip](https://github.com/komh/ksoftseq/files/12567564/uniaud.zip)

I've downloaded them from netlabs(http://trac.netlabs.org/uniaud).

**Comment by lerdmann** [comment]
> Oooops... This is because ksoftseq initializes kai only once at DLL load time which determines a default audio device. I'll try to fix this.
> 
> BTW, would you mind opening a separate ticket for this ?

Ok, see new ticket #12.

**Comment by lerdmann** [comment]
Hi komh,

I can confirm that with the UNIAUD drivers you have attached here that ksoftseq plays just fine with UNIAUD and that volume changes work without any hiccup.
So: it is a problem with UNIAUD.

The difference is UNIAUD32.SYS:

your version: 
Build Level Display Facility Version 6.12.675 Sep 25 2001
(C) Copyright IBM Corporation 1993-2001
Signature:       @#D Azarewicz:2.02.06#@##1## 6 May 2020 14:21:32      DAZAR1    :1.0.24:::06::SVN609@@Universal Audio Driver for OS/2 (c) 2020 D Azarewicz
Vendor:          D Azarewicz
Revision:        2.02.06
Date/Time:       6 May 2020 14:21:32    
Build Machine:   DAZAR1    
ASD Feature ID:  1.0.24
FixPak Version:  SVN609
File Version:    2.2.6
Description:     Universal Audio Driver for OS/2 (c) 2020 D Azarewicz


my version:
Build Level Display Facility Version 6.12.675 Sep 25 2001
(C) Copyright IBM Corporation 1993-2001
Signature:       @#D Azarewicz:3.01.02#@##1## 24 Jul 2021 10:01:23     DAZAR1    :5.10.50:::02::SVN689@@Universal Audio Driver for OS/2 (c) 2021 D Azarewicz
Vendor:          D Azarewicz
Revision:        3.01.02
Date/Time:       24 Jul 2021 10:01:23   
Build Machine:   DAZAR1    
ASD Feature ID:  5.10.50
FixPak Version:  SVN689
File Version:    3.1.2
Description:     Universal Audio Driver for OS/2 (c) 2021 D Azarewicz

**Comment by komh** [comment]
Ok, then I'll close this ticket as invalid. 

Thanks!
